apis/v1alpha1/adminnetworkpolicy_types.go

```diff
@@ -0,0 +1,174 @@
+	// ANPs with no egress rules do not affect egress traffic.
+	// +optional
+	// +kubebuilder:validation:MaxItems=100
+	Egress []AdminNetworkPolicyEgressRule `json:"egress,omitempty"`
+}
+
+// AdminNetworkPolicyIngressRule describes an action to take on a particular
+// set of traffic destined for pods selected by an AdminNetworkPolicy's
+// Subject field. The traffic must match both ports and from.
+type AdminNetworkPolicyIngressRule struct {
+	// Name is an identifier for this rule, that may be no more than 100 characters
+	// in length. This field should be used by the implementation to help
+	// improve observability, readability and error-reporting for any applied
+	// AdminNetworkPolicies.
+	// +optional
+	// +kubebuilder:validation:MaxLength=100
+	Name string `json:"name,omitempty"`
+
+	// Action specifies the affect this rule will have on matching traffic,
+	// currently the following actions are supported:
+	// Allow: allows the selected traffic
+	// Deny: denies the selected traffic
+	// Pass: instructs the selected traffic to skip any remaining ANP rules, and
+	// then pass execution to any NetworkPolicies that select the pod.
+	// If the pod is not selected by any NetworkPolicies then execution
+	// is passed to any BaselineAdminNetworkPolicies that select the pod.
+	// This field is mandatory.
+	Action AdminNetworkPolicyRuleAction `json:"action"`
+
+	// Ports allows for matching traffic based on port and protocols.
+	// If Ports is not set then traffic is not filtered via port.
+	// +optional
+	// +kubebuilder:validation:MaxItems=100
+	Ports []AdminNetworkPolicyPort `json:"ports,omitempty"`
+
+	// List of sources whose traffic this AdminNetworkPolicyRule applies to.
+	// If any adminNetworkPolicyPeer matches the source of incoming
+	// traffic then the specified action is applied.
+	// This field must be defined and contain at least one item.
+	// +kubebuilder:validation:MinItems=1
+	// +kubebuilder:validation:MaxItems=100
+	From []AdminNetworkPolicyPeer `json:"from"`
+}
+
+// AdminNetworkPolicyEgressRule describes an action to take on a particular
+// set of traffic originating from pods selected by a AdminNetworkPolicy's
+// Subject field. The traffic must match both ports and to.
+type AdminNetworkPolicyEgressRule struct {
+	// Name is an identifier for this rule, that may be no more than 100 characters
+	// in length. This field should be used by the implementation to help
+	// improve observability, readability and error-reporting for any applied
+	// AdminNetworkPolicies.
+	// +optional
+	// +kubebuilder:validation:MaxLength=100
+	Name string `json:"name,omitempty"`
+
+	// Action specifies the affect this rule will have on matching traffic,
+	// currently the following actions are supported:
+	// Allow: allows the selected traffic (even if it would otherwise have been denied by NetworkPolicy)
+	// Deny: denies the selected traffic (even if it would otherwise have been denied by NetworkPolicy)
```

Comment:
that should be "would otherwise have been _allowed_" but that feels less necessary for the "Deny" action? Denying is something that the implementation can just do immediately and then forget about, while "Allow" requires it to make other changes to the datapath (bypassing NP enforcement) so it needs the extra clarification